juju/entrypoint.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python3
+
+import os
+import sys
+
+# use modules from the lib directory - works under an assumption that
+# the current directory is the top-level charm directory ($JUJU_CHARM_DIR)
+sys.path.append('lib')
+
+from juju.framework import Framework # noqa
+from charm import Charm # noqa
+
+
+# reuse the same db file as used by charm-helpers
+DB_PATH = os.path.join(
+    os.environ.get('CHARM_DIR', ''), '.unit-state.db')
+
+try:
+    from charmhelpers.core.hookenv import log
+except (ImportError, ModuleNotFoundError):
+    log = print
+
+
+def emit_charm_event(charm, event_name):
+    """Emits a charm event based on a Juju event name
+
+    :param charm: A charm instance to emit an event from
+    :type charm: py:class:`juju.charm.Charm`
+    :param str event_name: A Juju event name to emit on a charm
+    """
+    formatted_event_name = event_name.replace('-', '_')
+    try:
+        event_to_emit = getattr(charm.on, formatted_event_name)
+    except AttributeError as e:
+        msg = f"Event {formatted_event_name} not defined for {charm}"
+        raise NotImplementedError(msg) from e
+
+    log(
+        f'Emitting a charm event {event_to_emit},'
+        f' based on a Juju event {event_name},'
+        f' event type {event_to_emit.event_type},'
+        f' event kind {event_to_emit.event_kind},'
+        f' event target {charm}'
+    )
+    event_to_emit.emit()
+
+
+def entrypoint():
+    """An entry point for a charm.
+
+    Sets up in-memory state (object hierarchy and observation etc.)
+
+    The expected hierarchy is a single charm object under the framework
+    object with many child objects under the charm object:
+
+    framework -> charm object -> member child objects (endpoints, components)
+
+    The charm class should hold all necessary setup code for the modeled
+    application such as object instantiation, setup of observation of events
+    (which may also be done in child classes).
+
+    """
+
+    # TODO: if Juju uniter agent crashes after exit(0) from the charm code
+    # the framework will commit the snapshot but Juju will not commit its
+    # operation
+    framework = Framework(data_path=DB_PATH)
+
+    # set up charm-specific in-memory state
+    charm = Charm(framework)
+
+    framework.reemit()
+
+    # process the Juju event relevant to the current hook execution
+    # JUJU_HOOK_NAME or JUJU_ACTION_NAME are not used to support simulation of
+    # events from debugging sessions.
+    juju_event_name = os.path.basename(sys.argv[0])
+    emit_charm_event(charm, juju_event_name)
+
+    framework.commit()
+    framework.close()
```

Comment:
The closing should be on a try/except, so we never leave this function without closing it. It's irrelevant if it's called as an entry point, but it's good practice to not expect this, as if we call this as a function and it fails we have a problem.